modules/gateway/helpers.go

```diff
@@ -36,10 +37,21 @@ func (g *Gateway) randomPeer() (network.Address, error) {
 	return "", ErrNoPeers
 }
 
-func (g *Gateway) broadcast(name string, arg, resp interface{}) {
+// threadedBroadcast broadcasts an RPC to all of the Gateway's peers. The
+// calls are run in parallel.
+func (g *Gateway) threadedBroadcast(name string, arg, resp interface{}) {
+	g.mu.RLock()
+	var wg sync.WaitGroup
+	wg.Add(len(g.peers))
 	for peer := range g.peers {
-		peer.RPC(name, arg, resp)
+		go func(peer network.Address) {
+			peer.RPC(name, arg, resp)
+			wg.Done()
+		}(peer)
 	}
+	// release lock while we wait for RPCs to complete
+	g.mu.RUnlock()
```

Comment:
This is still not good because you'll be holding the lock for at least as long as the timeout.

What you might need to do is add all of the peers into a slice.

```go
g.mu.RLock()
var peers []network.Address
for peer, _ := range g.peers {
    peers = append(peers, peer)
}
g.mu.RUnlock()
for _, peer := range peers {
    go func() {
        //
    }
}()
```